File: apps/mobile/lib/models/workflow.dart
enum NodeType {
  start,
  input,
  model,
  httpRequest,
  condition,
  template,
  end;

  String toJson() => name;
  static NodeType fromJson(String json) {
    return NodeType.values.firstWhere(
      (e) => e.name == json,
      orElse: () => NodeType.start,
    );
  }
}

class WorkflowNode {
  final String id;
  final String workflowId;
  final NodeType nodeType;
  final String name;
  final Map<String, dynamic> config;
  final double positionX;
  final double positionY;

  const WorkflowNode({
    required this.id,
    required this.workflowId,
    required this.nodeType,
    required this.name,
    this.config = const {},
    this.positionX = 0,
    this.positionY = 0,
  });

  factory WorkflowNode.fromJson(Map<String, dynamic> json) {
    return WorkflowNode(
      id: json['id'] as String,
      workflowId: json['workflow_id'] as String,
      nodeType: NodeType.fromJson(json['node_type'] as String),
      name: json['name'] as String,
      config: Map<String, dynamic>.from(json['config'] ?? {}),
      positionX: (json['position_x'] as num?)?.toDouble() ?? 0,
      positionY: (json['position_y'] as num?)?.toDouble() ?? 0,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'workflow_id': workflowId,
    'node_type': nodeType.toJson(),
    'name': name,
    'config': config,
    'position_x': positionX,
    'position_y': positionY,
  };
}

class WorkflowEdge {
  final String id;
  final String workflowId;
  final String sourceNodeId;
  final String targetNodeId;
  final String? sourceHandle;
  final String? targetHandle;
  final String? conditionExpression;

  const WorkflowEdge({
    required this.id,
    required this.workflowId,
    required this.sourceNodeId,
    required this.targetNodeId,
    this.sourceHandle,
    this.targetHandle,
    this.conditionExpression,
  });

  factory WorkflowEdge.fromJson(Map<String, dynamic> json) {
    return WorkflowEdge(
      id: json['id'] as String,
      workflowId: json['workflow_id'] as String,
      sourceNodeId: json['source_node_id'] as String,
      targetNodeId: json['target_node_id'] as String,
      sourceHandle: json['source_handle'] as String?,
      targetHandle: json['target_handle'] as String?,
      conditionExpression: json['condition_expression'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'workflow_id': workflowId,
    'source_node_id': sourceNodeId,
    'target_node_id': targetNodeId,
    'source_handle': sourceHandle,
    'target_handle': targetHandle,
    'condition_expression': conditionExpression,
  };
}

class Workflow {
  final String id;
  final String agentId;
  final String name;
  final String? description;
  final List<WorkflowNode> nodes;
  final List<WorkflowEdge> edges;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Workflow({
    required this.id,
    required this.agentId,
    required this.name,
    this.description,
    this.nodes = const [],
    this.edges = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory Workflow.fromJson(Map<String, dynamic> json) {
    return Workflow(
      id: json['id'] as String,
      agentId: json['agent_id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      nodes: (json['nodes'] as List?)
          ?.map((n) => WorkflowNode.fromJson(n))
          .toList() ?? [],
      edges: (json['edges'] as List?)
          ?.map((e) => WorkflowEdge.fromJson(e))
          .toList() ?? [],
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'agent_id': agentId,
    'name': name,
    'description': description,
    'nodes': nodes.map((n) => n.toJson()).toList(),
    'edges': edges.map((e) => e.toJson()).toList(),
    'created_at': createdAt.toIso8601String(),
    'updated_at': updatedAt.toIso8601String(),
  };
}

class WorkflowInput {
  final String agentId;
  final String name;
  final String? description;
  final List<WorkflowNodeInput> nodes;
  final List<WorkflowEdgeInput> edges;

  const WorkflowInput({
    required this.agentId,
    required this.name,
    this.description,
    this.nodes = const [],
    this.edges = const [],
  });

  Map<String, dynamic> toJson() => {
    'agent_id': agentId,
    'name': name,
    'description': description,
    'nodes': nodes.map((n) => n.toJson()).toList(),
    'edges': edges.map((e) => e.toJson()).toList(),
  };
}

class WorkflowNodeInput {
  final NodeType nodeType;
  final String name;
  final Map<String, dynamic> config;
  final double positionX;
  final double positionY;

  const WorkflowNodeInput({
    required this.nodeType,
    required this.name,
    this.config = const {},
    this.positionX = 0,
    this.positionY = 0,
  });

  Map<String, dynamic> toJson() => {
    'node_type': nodeType.toJson(),
    'name': name,
    'config': config,
    'position_x': positionX,
    'position_y': positionY,
  };
}

class WorkflowEdgeInput {
  final String sourceNodeId;
  final String targetNodeId;
  final String? sourceHandle;
  final String? targetHandle;
  final String? conditionExpression;

  const WorkflowEdgeInput({
    required this.sourceNodeId,
    required this.targetNodeId,
    this.sourceHandle,
    this.targetHandle,
    this.conditionExpression,
  });

  Map<String, dynamic> toJson() => {
    'source_node_id': sourceNodeId,
    'target_node_id': targetNodeId,
    'source_handle': sourceHandle,
    'target_handle': targetHandle,
    'condition_expression': conditionExpression,
  };
}

File: crates/core/src/models/workflow.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;

use crate::types::NodeType;

/// A directed acyclic graph (DAG) workflow definition.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Workflow {
    pub id: Uuid,
    pub agent_id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub nodes: Vec<WorkflowNode>,
    pub edges: Vec<WorkflowEdge>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// A single node in a workflow.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowNode {
    pub id: Uuid,
    pub workflow_id: Uuid,
    pub node_type: NodeType,
    pub name: String,
    /// Node-specific configuration as JSON.
    pub config: Value,
    pub position_x: f64,
    pub position_y: f64,
}

/// A directed edge connecting two nodes in a workflow.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowEdge {
    pub id: Uuid,
    pub workflow_id: Uuid,
    pub source_node_id: Uuid,
    pub target_node_id: Uuid,
    pub source_handle: Option<String>,
    pub target_handle: Option<String>,
    /// Optional condition expression for conditional branching.
    pub condition_expression: Option<String>,
}

/// Status of a workflow run.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RunStatus {
    Pending,
    Running,
    Paused,
    Failed,
    Succeeded,
    Cancelled,
    TimedOut,
}

/// Status of a single node execution.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum NodeStatus {
    Pending,
    Running,
    Paused,
    WaitingApproval,
    Failed,
    Succeeded,
    Cancelled,
}

/// A single execution run of a workflow.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowRun {
    pub id: Uuid,
    pub workflow_id: Uuid,
    pub status: RunStatus,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub error: Option<String>,
    pub trigger_type: String,
    pub created_at: DateTime<Utc>,
}

/// Execution record of a single node within a workflow run.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NodeRun {
    pub id: Uuid,
    pub workflow_run_id: Uuid,
    pub node_id: Uuid,
    pub node_type: NodeType,
    pub status: NodeStatus,
    pub input: Option<Value>,
    pub output: Option<Value>,
    pub error: Option<String>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub attempts: i32,
    pub max_attempts: i32,
    pub created_at: DateTime<Utc>,
}

impl Workflow {
    /// Check if the workflow forms a valid DAG (no cycles).
    /// Uses Kahn's algorithm for topological sort.
    pub fn is_valid_dag(&self) -> bool {
        let mut in_degree: std::collections::HashMap<Uuid, usize> =
            self.nodes.iter().map(|n| (n.id, 0)).collect();

        let mut adjacency: std::collections::HashMap<Uuid, Vec<Uuid>> =
            self.nodes.iter().map(|n| (n.id, Vec::new())).collect();

        for edge in &self.edges {
            if let Some(neighbors) = adjacency.get_mut(&edge.source_node_id) {
                neighbors.push(edge.target_node_id);
            }
            if let Some(degree) = in_degree.get_mut(&edge.target_node_id) {
                *degree += 1;
            }
        }

        let mut queue: Vec<Uuid> = in_degree
            .iter()
            .filter(|&(_, deg)| *deg == 0)
            .map(|(&id, _)| id)
            .collect();

        let mut visited = 0;
        while let Some(node) = queue.pop() {
            visited += 1;
            if let Some(neighbors) = adjacency.get(&node) {
                for &next in neighbors {
                    if let Some(degree) = in_degree.get_mut(&next) {
                        *degree -= 1;
                        if *degree == 0 {
                            queue.push(next);
                        }
                    }
                }
            }
        }

        visited == self.nodes.len()
    }

    /// Get the start node of the workflow.
    pub fn start_node(&self) -> Option<&WorkflowNode> {
        self.nodes.iter().find(|n| n.node_type == NodeType::Start)
    }

    /// Get the end nodes of the workflow.
    pub fn end_nodes(&self) -> Vec<&WorkflowNode> {
        self.nodes
            .iter()
            .filter(|n| n.node_type == NodeType::End)
            .collect()
    }

    /// Get downstream nodes of a given node.
    pub fn downstream_nodes(&self, node_id: Uuid) -> Vec<Uuid> {
        self.edges
            .iter()
            .filter(|e| e.source_node_id == node_id)
            .map(|e| e.target_node_id)
            .collect()
    }
}

impl WorkflowRun {
    /// Check if the run is in a terminal state.
    pub fn is_terminal(&self) -> bool {
        matches!(
            self.status,
            RunStatus::Succeeded | RunStatus::Failed | RunStatus::Cancelled | RunStatus::TimedOut
        )
    }

    /// Check if the run can be retried.
    pub fn can_retry(&self) -> bool {
        matches!(self.status, RunStatus::Failed | RunStatus::TimedOut)
    }
}

impl NodeRun {
    /// Check if this node run is in a terminal state.
    pub fn is_terminal(&self) -> bool {
        matches!(
            self.status,
            NodeStatus::Succeeded | NodeStatus::Failed | NodeStatus::Cancelled
        )
    }
}
